File: python/ndsq/src/ndsq/values.py
"""Index values (finite or +/-inf) and bounds with the [n]-bracket implicit convention."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal, Union

from .errors import NdsqError, Reason

# A single index coordinate: a finite int, or the infinity sentinels.
IndexValue = Union[int, Literal["-inf", "+inf"]]


def parse_index_value(raw: object) -> IndexValue:
    """Validate a JSON scalar as an IndexValue. Rejects bool (a Python int subclass)."""
    if isinstance(raw, bool):
        raise NdsqError(Reason.INVALID_JSON, f"index value must be an integer, got bool {raw!r}")
    if isinstance(raw, int):
        return raw
    if raw == "-inf" or raw == "+inf":
        return raw  # type: ignore[return-value]
    raise NdsqError(Reason.INVALID_JSON, f"invalid index value: {raw!r}")


@dataclass(frozen=True)
class ImplicitValue:
    """An index bound plus an explicit/implicit flag.

    JSON: a bare value is explicit (`7`, `"-inf"`); the same wrapped in a
    one-element array is implicit (`[7]`, `["-inf"]`).
    """

    value: IndexValue
    implicit: bool

    @classmethod
    def explicit(cls, value: IndexValue) -> "ImplicitValue":
        return cls(value=value, implicit=False)

    @classmethod
    def from_json(cls, raw: object) -> "ImplicitValue":
        if isinstance(raw, list):
            if len(raw) != 1:
                raise NdsqError(Reason.INVALID_JSON, "implicit bound must be a 1-element array")
            return cls(value=parse_index_value(raw[0]), implicit=True)
        return cls(value=parse_index_value(raw), implicit=False)

    def to_json(self) -> object:
        return [self.value] if self.implicit else self.value

File: typescript/src/values.ts
import { NdselError, Reason } from "./errors.ts";

/**
 * An index integer. Values within the JS safe-integer range are `number`;
 * values outside it (but within i64) are `bigint` (§3.5). The canonical form is
 * representation-independent: a value is `number` iff it is a safe integer.
 */
export type Int = number | bigint;

/** A single index coordinate: an integer, or an infinity sentinel. */
export type IndexValue = Int | "-inf" | "+inf";

/** A JSON-level bound: a value, or that value wrapped in a 1-element array (implicit). */
export type BoundJson = IndexValue | [IndexValue];

/** Internal parsed bound: the value plus its explicit/implicit flag. */
export interface ParsedBound {
  value: IndexValue;
  implicit: boolean;
}

// 64-bit signed range (the canonical contract, §3.5) as bigints.
const I64_MIN = -(2n ** 63n);
const I64_MAX = 2n ** 63n - 1n;
const SAFE_MAX = BigInt(Number.MAX_SAFE_INTEGER); // 2^53 - 1

/** Promote an `Int` to `bigint` for arithmetic. */
export function toBig(x: Int): bigint {
  return typeof x === "bigint" ? x : BigInt(x);
}

/** Demote a `bigint` to `number` when it is a safe integer; keep `bigint` otherwise. */
export function demote(x: bigint): Int {
  return x >= -SAFE_MAX && x <= SAFE_MAX ? Number(x) : x;
}

function checkI64(x: bigint, what: string): Int {
  if (x < I64_MIN || x > I64_MAX) {
    throw new NdselError(Reason.InvalidJson, `${what} value ${x} is out of 64-bit signed range`);
  }
  return demote(x);
}

export function parseIndexValue(raw: unknown): IndexValue {
  if (typeof raw === "bigint") return checkI64(raw, "index");
  if (typeof raw === "number" && Number.isSafeInteger(raw)) return raw;
  if (raw === "-inf" || raw === "+inf") return raw;
  throw new NdselError(Reason.InvalidJson, `invalid index value: ${JSON.stringify(raw)}`);
}

export function parseBound(raw: unknown): ParsedBound {
  if (Array.isArray(raw)) {
    if (raw.length !== 1) {
      throw new NdselError(Reason.InvalidJson, "implicit bound must be a 1-element array");
    }
    return { value: parseIndexValue(raw[0]), implicit: true };
  }
  return { value: parseIndexValue(raw), implicit: false };
}

export function boundToJSON(b: ParsedBound): BoundJson {
  return b.implicit ? [b.value] : b.value;
}

/** Validate a plain integer in 64-bit signed range (no sentinels). */
export function requireInt(raw: unknown, what: string): Int {
  if (typeof raw === "bigint") return checkI64(raw, what);
  if (typeof raw === "number" && Number.isSafeInteger(raw)) return raw;
  throw new NdselError(Reason.InvalidJson, `${what} must be an integer, got ${JSON.stringify(raw)}`);
}

export function requireArray(raw: unknown, what: string): unknown[] {
  if (Array.isArray(raw)) return raw;
  throw new NdselError(Reason.InvalidJson, `${what} must be an array, got ${JSON.stringify(raw)}`);
}

export function requireIntArray(raw: unknown, what: string): Int[] {
  return requireArray(raw, what).map((v, i) => requireInt(v, `${what}[${i}]`));
}

export function requireStringArray(raw: unknown, what: string): string[] {
  return requireArray(raw, what).map((v, i) => {
    if (typeof v !== "string") {
      throw new NdselError(Reason.InvalidJson, `${what}[${i}] must be a string, got ${JSON.stringify(v)}`);
    }
    return v;
  });
}
